src/generators/modelsGenerator.ts

```diff
@@ -16,31 +37,332 @@ export function generateModels(clientDetails: ClientDetails, project: Project) {
+ */
+function writeUniontypes({ objects }: ClientDetails, modelsFile: SourceFile) {
+  objects
+    .filter(
+      obj => obj.kind === ObjectKind.Polymorphic && obj.children.length > 0
+    )
+    .forEach(obj => {
+      const polymorphicObject = obj as PolymorphicObjectDetails;
+      const childrenNames = [
+        obj.name,
+        ...polymorphicObject.children.map(c => {
+          return c.schema.children && c.schema.children.immediate.length
+            ? `${c.name}Union`
+            : c.name;
+        })
+      ];
+      modelsFile.addTypeAlias({
+        name: `${obj.name}Union`,
+        isExported: true,
+        type: childrenNames.join(" | "),
+        trailingTrivia: writer => writer.newLine()
+      });
     });
+}
+
+function writeOptionalParameters(
+  operationGroupName: string,
+  operationName: string,
+  optionalParams: ParameterDetails[],
+  modelsIndexFile: SourceFile
+) {
+  if (!optionalParams || !optionalParams.length) {
+    return;
   }
+
+  modelsIndexFile.addInterface({
+    name: `${operationGroupName}${operationName}OptionalParams`,
+    docs: ["Optional parameters."],
+    isExported: true,
+    extends: ["coreHttp.RequestOptionsBase"],
+    properties: optionalParams.map<PropertySignatureStructure>(p => ({
+      name: p.name,
+      hasQuestionToken: !p.required,
+      type: p.typeDetails.typeName || "TODO NO TYPE",
+      docs: p.description ? [p.description] : undefined,
+      kind: StructureKind.PropertySignature
+    }))
+  });
 }
+
+/**
+ * Extracts all properties from ObjectDetails and returns a list of PropertySignatureStructure
+ * @param objectDetails Object description
+ */
+function getProperties(
+  objectDetails: ObjectDetails
+): PropertySignatureStructure[] {
+  const { properties } = objectDetails;
+  const getTypename = (property: PropertyDetails) =>
+    property.name === "siblings"
+      ? `${(objectDetails as PolymorphicObjectDetails).unionName}[]`
+      : property.type;
+
+  return properties
+    .filter(property => !property.isConstant && !property.isDiscriminator)
+    .map<PropertySignatureStructure>(property => ({
+      name: property.name,
+      hasQuestionToken: !property.required,
+      isReadonly: property.readOnly,
+      type: getTypename(property),
+      docs: property.description ? [property.description] : undefined,
+      kind: StructureKind.PropertySignature
+    }));
+}
+
+/**
+ * This function enahnces a list of PropertySignatures with the Polymorphic discriminator property if needed
+ * @param model ObjectDetails
+ * @param properties Properties to enhance
+ */
+function withDiscriminator(
+  model: ObjectDetails,
+  properties: PropertySignatureStructure[]
+): PropertySignatureStructure[] {
+  const discriminator = (model as PolymorphicObjectDetails).discriminator;
+  if (!discriminator) {
+    return properties;
+  }
+
+  const discProps = keys(discriminator).map<PropertySignatureStructure>(
+    key => ({
+      docs: [`Polymorphic discriminator`],

```

Comment:
I suppose a discriminator doesn't specify its own documentation string?  I don't know what the model does in this case, but seems like it'd have to be a normal property on the object which would have some meaning to the user

Reply:
In the current samples the description field is empty. I'll dig a bit more and probably we can honor the discriminator description if present or have this as fallback